File: dp/house_robber_ii.rb
# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
    return 0 if nums.empty?
    return nums.max if nums.size <= 2

    [rob_seq(nums[0, nums.size - 1]), rob_seq(nums[1, nums.size])].max
end

def rob_seq(nums)
    return 0 if nums.empty?
    return nums.max if nums.size <= 2

    n = nums.size
    dp = [-1] * n
    dp[0], dp[1] = nums[0], [nums[0], nums[1]].max

    2.upto(n - 1) do |i|
        dp[i] = [dp[i - 1], dp[i - 2] + nums[i]].max
    end

    dp[n - 1]
end

File: dp/house_robber_ii.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums: return 0
        if len(nums) <= 2: return max(nums)

        return max(self.rob_flat(nums[:-1]), self.rob_flat(nums[1:]))

    def rob_flat(self, nums: List[int]) -> int:
        if not nums: return 0
        if len(nums) <= 2: return max(nums)

        n = len(nums)
        dp = [-1] * n
        dp[0], dp[1] = nums[0], max(nums[0], nums[1])

        for i in range(2, n):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])

        return dp[n - 1]


# 213. House Robber II
# https://leetcode.com/problems/house-robber-ii/description/


sol = Solution()

assert sol.rob([]) == 0
assert sol.rob([0, 0, 0]) == 0
assert sol.rob([1, 1, 1]) == 1
assert sol.rob([1, 1]) == 1
assert sol.rob([2, 3, 2]) == 3
assert sol.rob([1, 2, 3, 1]) == 4
